controller/object_store_controller.go

```diff
@@ -0,0 +1,976 @@
+							Name:  "s3gw",
+							Image: osc.getObjectStoreImage(store),
+							Args: []string{
+								"--rgw-dns-name", (*osc.getLocalEndpointURL(store)).DomainName,
+								"--rgw-backend-store", "sfs",
+								"--rgw_frontends", fmt.Sprintf(
+									"beast port=%d, status port=%d",
+									types.ObjectStoreContainerPort,
+									types.ObjectStoreStatusContainerPort,
+								),
+							},
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "s3",
+									ContainerPort: types.ObjectStoreContainerPort,
+									Protocol:      "TCP",
+								},
+								{
+									Name:          "status",
+									ContainerPort: types.ObjectStoreStatusContainerPort,
+									Protocol:      "TCP",
+								},
+							},
+							EnvFrom: []corev1.EnvFromSource{
+								{
+									SecretRef: &corev1.SecretEnvSource{
+										LocalObjectReference: corev1.LocalObjectReference{
+											Name: store.Spec.Credentials.Name,
+										},
+									},
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      genVolumeMountName(store),
+									MountPath: "/data",
+								},
+							},
+						},
+						{
+							Name:  "s3gw-ui",
+							Image: osc.getObjectStoreUIImage(store),
+							Args:  []string{},
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "ui",
+									ContainerPort: types.ObjectStoreUIContainerPort,
+									Protocol:      "TCP",
+								},
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name: "S3GW_SERVICE_URL",
+									Value: fmt.Sprintf("http://%v/",
+										(*osc.getLocalEndpointURL(store)).DomainName),
+								},
+								{
+									Name:  "S3GW_UI_LOCATION",
+									Value: fmt.Sprintf("/%v", store.Name),
+								},
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: genVolumeMountName(store),
+							VolumeSource: corev1.VolumeSource{
+								PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+									ClaimName: genPVCName(store),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return osc.ds.CreateDeployment(&dpl)
+}
+
+// getLocalEndpointURL generates the URL of the namespace-local endpoint for the
+// object store. This endpoint is implicit
+func (osc *ObjectStoreController) getLocalEndpointURL(store *longhorn.ObjectStore) *longhorn.ObjectStoreEndpointSpec {
+	return &longhorn.ObjectStoreEndpointSpec{
+		Name:       "local",
+		DomainName: fmt.Sprintf("%s.%s.svc", store.Name, osc.namespace),
+	}
+}
+
+func (osc *ObjectStoreController) isResponsibleFor(store *longhorn.ObjectStore) bool {
+	isResponsible := osc.controllerID == store.Status.OwnerID
+	vol, err := osc.ds.GetVolume(genPVName(store))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to find volume for object store %v: %v", store.Name, err))
+		return true
+	}
+	needsNewOwner := store.Status.OwnerID != vol.Status.OwnerID
+	return isResponsible || needsNewOwner
+}
```

Comment:
Because we are tieing the owner of the ObjectStore to be the owner of the volume (which is a good idea). I think we don't need the field `store.Status.OwnerID`. Suggesting to remove that field and modify this function to:

```
func (osc *ObjectStoreController) isResponsibleFor(store *longhorn.ObjectStore) (bool, error) {
	var err error
	defer func() {
		err = errors.Wrap(err, "error while checking isResponsibleFor")
	}()

	volume, err := osc.ds.GetVolumeRO(snap.Spec.Volume)
	if err != nil {
		if apierrors.IsNotFound(err) {
			return true, nil
		}
		return false, err
	}
	return osc.controllerID == volume.Status.OwnerID, nil
}
```